File: cmd/audit.go
package cmd

import (
	"fmt"
	"os"

	"github.com/envsync/internal/parser"
	"github.com/fatih/color"
	"github.com/spf13/cobra"
)

var auditCmd = &cobra.Command{
	Use:   "audit",
	Short: "Audit a single environment for missing/undefined keys",
	Long: `Audit compares an environment against the .env.example source of truth.
Useful as a pre-deployment check in CI/CD pipelines.

Examples:
  envsync audit --env staging
  envsync audit --env production --fail-on-missing
  envsync audit --env dev --threshold 10`,
	Run: runAudit,
}

var (
	auditEnv          string
	auditFailOnMissing bool
	auditThreshold    int
)

func init() {
	auditCmd.Flags().StringVar(&auditEnv, "env", "dev", "Environment to audit")
	auditCmd.Flags().BoolVar(&auditFailOnMissing, "fail-on-missing", false, "Exit with code 1 if any keys are missing")
	auditCmd.Flags().IntVar(&auditThreshold, "threshold", 0, "Max allowed drift count (0 = unlimited)")
}

func runAudit(cmd *cobra.Command, args []string) {
	configPath, _ := cmd.Root().PersistentFlags().GetString("config")

	printBanner(fmt.Sprintf("Audit: %s", auditEnv))

	cfg, err := parser.LoadConfig(configPath)
	if err != nil {
		exitErr("Failed to load config", err)
	}

	env, err := parser.LoadEnvironment(cfg, auditEnv)
	if err != nil {
		exitErr(fmt.Sprintf("Failed to load environment '%s'", auditEnv), err)
	}

	truth, err := parser.LoadSourceOfTruth(cfg)
	if err != nil {
		exitErr("Failed to load source of truth (.env.example)", err)
	}

	missing := []string{}
	empty := []string{}
	extra := []string{}
	ok := 0

	for key := range truth {
		val, exists := env[key]
		if !exists {
			missing = append(missing, key)
		} else if val == "" || val == "CHANGE_ME" || val == "TODO" {
			empty = append(empty, key)
		} else {
			ok++
		}
	}

	for key := range env {
		if _, exists := truth[key]; !exists {
			extra = append(extra, key)
		}
	}

	// Print results
	if len(missing) > 0 {
		color.Red("\n✗ MISSING keys (not in %s):\n", auditEnv)
		for _, k := range missing {
			color.Red("   - %s\n", k)
		}
	}

	if len(empty) > 0 {
		color.Yellow("\n⚠ EMPTY / PLACEHOLDER keys:\n")
		for _, k := range empty {
			color.Yellow("   ~ %s\n", k)
		}
	}

	if len(extra) > 0 {
		color.Blue("\nℹ EXTRA keys (not in .env.example):\n")
		for _, k := range extra {
			color.Blue("   + %s\n", k)
		}
	}

	fmt.Println()
	color.Green("✔ OK: %d keys\n", ok)
	color.Red("✗ Missing: %d keys\n", len(missing))
	color.Yellow("~ Empty: %d keys\n", len(empty))
	color.Blue("+ Extra: %d keys\n", len(extra))

	totalDrift := len(missing) + len(empty)

	if auditThreshold > 0 && totalDrift > auditThreshold {
		color.Red("\n✗ AUDIT FAILED: Drift count %d exceeds threshold %d\n", totalDrift, auditThreshold)
		os.Exit(1)
	}

	if auditFailOnMissing && len(missing) > 0 {
		color.Red("\n✗ AUDIT FAILED: Missing keys detected\n")
		os.Exit(1)
	}

	if totalDrift == 0 {
		success(fmt.Sprintf("Environment '%s' is clean ✓", auditEnv))
	} else {
		warn(fmt.Sprintf("Environment '%s' has %d issues", auditEnv, totalDrift))
	}
}

File: cmd/audit.py
import sys
import click
from colorama import Fore
from core.parser import load_config, load_environment, load_source_of_truth
from commands.root import cli, print_banner


@cli.command()
@click.option("--env", default="dev", help="Environment to audit")
@click.option("--fail-on-missing", is_flag=True, help="Exit 1 if any keys missing")
@click.option("--threshold", default=0, help="Max allowed drift count")
@click.pass_context
def audit(ctx, env, fail_on_missing, threshold):
    """Audit a single environment for missing or undefined keys"""

    config_path = ctx.obj["config_path"]
    print_banner(f"Audit: {env}")

    try:
        config = load_config(config_path)
    except Exception as e:
        print(Fore.RED + f"✗ Failed to load config: {e}")
        sys.exit(1)

    try:
        environment = load_environment(config, env)
    except Exception as e:
        print(Fore.RED + f"✗ Failed to load environment '{env}': {e}")
        sys.exit(1)

    try:
        truth = load_source_of_truth(config)
    except Exception as e:
        print(Fore.RED + f"✗ Failed to load source of truth: {e}")
        sys.exit(1)

    missing = []
    empty = []
    extra = []
    ok = 0

    for key in truth:
        if key not in environment:
            missing.append(key)
        elif not environment[key] or environment[key] in ["CHANGE_ME", "TODO"]:
            empty.append(key)
        else:
            ok += 1

    for key in environment:
        if key not in truth:
            extra.append(key)

    if missing:
        print(Fore.RED + f"\n✗ MISSING keys (not in {env}):")
        for k in missing:
            print(Fore.RED + f"   - {k}")

    if empty:
        print(Fore.YELLOW + "\n⚠ EMPTY / PLACEHOLDER keys:")
        for k in empty:
            print(Fore.YELLOW + f"   ~ {k}")

    if extra:
        print(Fore.BLUE + "\nℹ EXTRA keys (not in .env.example):")
        for k in extra:
            print(Fore.BLUE + f"   + {k}")

    print()
    print(Fore.GREEN + f"✔ OK: {ok} keys")
    print(Fore.RED + f"✗ Missing: {len(missing)} keys")
    print(Fore.YELLOW + f"~ Empty: {len(empty)} keys")
    print(Fore.BLUE + f"+ Extra: {len(extra)} keys")

    total_drift = len(missing) + len(empty)

    if threshold > 0 and total_drift > threshold:
        print(Fore.RED + f"\n✗ AUDIT FAILED: Drift count {total_drift} exceeds threshold {threshold}")
        sys.exit(1)

    if fail_on_missing and missing:
        print(Fore.RED + "\n✗ AUDIT FAILED: Missing keys detected")
        sys.exit(1)

    if total_drift == 0:
        print(Fore.GREEN + f"\n✔ Environment '{env}' is clean ✓")
    else:
        print(Fore.YELLOW + f"\n⚠ Environment '{env}' has {total_drift} issues")
